Blackjack

luyaosuperman opened the pull request:
This is the package of blackjack.
Implemented
(1) one round of game only, for the moment
(2) multiple decks.
(3) multiple users
(4) hit, stand, double, split, surrender

Lack of
(1) Chips
(2) Multiple rounds

files:

./__init__.py
./card.py
./dealer.py
./game.py <--- "Game" class implements the logic
./player.py
./xprint.py

./test.sh
./tests/cardTest.py
./tests/dealerTest.py
./tests/gameTest.py
./tests/__init__.py
./tests/playerTest.py

![umlattempt](https://user-images.githubusercontent.com/2969220/27565899-a09b6398-5b24-11e7-833b-207568c3fb00.PNG)
![umlattempt1](https://user-images.githubusercontent.com/2969220/27565901-a22fb7ae-5b24-11e7-9867-b3bd87389e80.PNG)


+++++
Most of the time was spent on implementing test cases. I have put it into a package and utilizes the "unittest discover" to automate it a bit. I browsed some projects in Github to get some idea about test organziation.

However
(1) it is still very hard to test class "Game" in game.py as it is basically manipulating other classes without returning any value. I am not sure how to design the unittest for it. Or, how should I rewrite it to make it more test friendly.

(2) I tried to seperate "Player" and "Dealer" as much as possible while I have to pass one of the func pointer "Dealer.dealACard" into Player, so that player can get a card from dealer. I am not very sure if I can do better in seperating two classes.

(3) cursesTest.py was deleted locally and it won't apper in git fetch/pull. I am still looking into how to remove it from Github repo.

luyaosuperman committed:
commit

luyaosuperman committed:
commit

luyaosuperman committed:
commity

Commit:
commit

luyaosuperman committed:
commit

luyaosuperman committed:
commit

luyaosuperman committed:
commit

Commit:
commit

cjwfuller commented on homework/blackjack/card.py:
```diff
@@ -0,0 +1,64 @@
+from xprint import xPrint, printStack
+
+
+class Card():
```
class Card(object): # read about "old and new classes" in https://docs.python.org/3/whatsnew/2.2.html

cjwfuller commented on homework/blackjack/card.py:
```diff
@@ -0,0 +1,64 @@
+        "SPADES",
+    }
+
+    # __usedCombination = set()
```
remove dead code

cjwfuller commented on homework/blackjack/card.py:
```diff
@@ -0,0 +1,64 @@
+    # It should contain the set of cards which is already shuffled,
+    # in form of tuple: {(Cards.CLUBS, 1), (Cards.SPADES, 2)}
+
+    def __init__(self, suit, rank):
```
Consider using docstrings

cjwfuller commented on homework/blackjack/card.py:
```diff
@@ -0,0 +1,64 @@
+            raise ValueError("the suit provided is not valid")
+        if not (self.rank >= 1 and self.rank <= 13):
+            raise ValueError("the rank provided is not valid")
+        # If, else, raise
```
you can actually drop the parentheses

cjwfuller commented on homework/blackjack/dealer.py:
```diff
@@ -0,0 +1,138 @@
+            self.cards[i], self.cards[j] = self.cards[j], self.cards[i]
+
+    @printStack
+    def dealACard(self):
```
dropping the 'A's might be clearer e.g. dealCard

cjwfuller commented on homework/blackjack/dealer.py:
```diff
@@ -0,0 +1,138 @@
+        )
+
+    @printStack
+    def act(self):
```
what does act mean?

cjwfuller commented on homework/blackjack/game.py:
```diff
@@ -0,0 +1,159 @@
+            )
+
+    @printStack
+    def __initGame(self):
```
if this is initialising, maybe it should be called from the constructor?

cjwfuller replied:
As discussed - no longer relevant (this looks ok)

cjwfuller commented on homework/blackjack/game.py:
```diff
@@ -0,0 +1,159 @@
+        """
+        for player in self.players:
+            if player.isPlayerFinished() == False:
+                return
```
this is actually doing `return None` - maybe be explicit and just return None

cjwfuller commented on homework/blackjack/game.py:
```diff
@@ -0,0 +1,159 @@
+        And end the game if necessary
+        """
+        for player in self.players:
+            if player.isPlayerFinished() == False:
```
I think `is None` would follow PEP8 more closely but check

cjwfuller replied:
Discussed, this is good

cjwfuller commented on homework/blackjack/game.py:
```diff
@@ -0,0 +1,159 @@
+            player.finalAction()
+            self.__isPlayerWin(player)
+
+            '''if self.__isPlayerWin(player):
```
dead code

cjwfuller commented on homework/blackjack/tests/dealerTest.py:
```diff
@@ -0,0 +1,124 @@
+        #xprint.debugPrint = True
+        xPrint(cardsDict)
+
+    def testCreateADeckOfCard(self):
```
Test that a deck of cards is valid e.g. 1 ace of spaces, 1 queen of hearts etc..?

luyaosuperman replied:
Added func __validateDeckCompleteness() in dealerTest to validate the completeness of a deck of cards.

cjwfuller reviewed (commented):
Good effort 👍  a few comments to look at.  Maybe take a pass through these and we can do another review!

luyaosuperman committed:
commit

luyaosuperman committed:
Merge branch 'master' into blackjack

cjwfuller commented:
@luyaosuperman see replies to these comments / questions below:

(1) it is still very hard to test class "Game" in game.py as it is basically manipulating other classes without returning any value. I am not sure how to design the unittest for it. Or, how should I rewrite it to make it more test friendly.

This file https://github.com/luyaosuperman/python3-playground/blob/blackjack/homework/blackjack/tests/gameTest.py ?

It's great that you have introduced a mock for Game.  You might want to try playing around with the @patch to see if you can simplify it to @patch('Game._Game__game', return_value=None) because Game is already defined.

Some test ideas.  That that:
- a player can win
- a player can lose
- __judge works as intended
- a game is finished when it has finished
- a game has not finished when it is still in progress
... that sort of thing

(2) I tried to seperate "Player" and "Dealer" as much as possible while I have to pass one of the func pointer "Dealer.dealACard" into Player, so that player can get a card from dealer. I am not very sure if I can do better in seperating two classes.

Great (Separation Of Concerns).  We discussed but I'll comment here too.  Instead of passing a reference to a function, you can just pass an object and use its public methods.  You shouldn't need to pass in a function.

(3) cursesTest.py was deleted locally and it won't apper in git fetch/pull. I am still looking into how to remove it from Github repo.

This file - https://github.com/luyaosuperman/python3-playground/blob/blackjack/homework/blackjack/cursesTest.py ?

You should be able to do something like:

cd blackjack
rm cursesTest.py # delete the file from your machine
git rm cursesTest.py # delete the file from git
git commit -m "remove cursesTest.py"
git push origin blackjack

luyaosuperman committed:
add test
testPlayerWin
testPlayerLose

split test test__judege to two, to better test out scenario that if game should finish or not
test__judge_GameFinish
test__judge_GameContinue

luyaosuperman commented:
@cjwfuller 
state of the player is added to the class Player to make it more testable.
On top of the state, two tests are added into gameTest
testPlayerWin()
testPlayerLose()
It tests if the player state will be set correctly.

Also, add two more tests regarding if game should be finished or not
test__judge_GameFinish()
test__judge_GameContinue()

Thank you!

luyaosuperman committed:
PEP8 format

cjwfuller commented:
This looks good.
